bindings/python/bergamot.cpp

```diff
@@ -0,0 +1,227 @@
+#include <translator/annotation.h>
+#include <translator/parser.h>
+#include <translator/project_version.h>
+#include <translator/response.h>
+#include <translator/response_options.h>
+#include <translator/service.h>
+#include <translator/translation_model.h>
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace py = pybind11;
+
+using marian::bergamot::AnnotatedText;
+using marian::bergamot::ByteRange;
+using marian::bergamot::ConcatStrategy;
+using marian::bergamot::Response;
+using marian::bergamot::ResponseOptions;
+using Service = marian::bergamot::AsyncService;
+using _Model = marian::bergamot::TranslationModel;
+using Model = std::shared_ptr<_Model>;
+using Alignment = std::vector<std::vector<float>>;
+using Alignments = std::vector<Alignment>;
+
+PYBIND11_MAKE_OPAQUE(std::vector<Response>);
+PYBIND11_MAKE_OPAQUE(std::vector<std::string>);
+PYBIND11_MAKE_OPAQUE(Alignments);
+
+class ServicePyAdapter {
+public:
+  ServicePyAdapter(const Service::Config &config)
+      : service_(make_service(config)) {}
+
+  std::shared_ptr<_Model> modelFromConfig(const std::string &config) {
+    auto parsedConfig = marian::bergamot::parseOptionsFromString(config);
+    return service_.createCompatibleModel(parsedConfig);
+  }
+
+  std::shared_ptr<_Model> modelFromConfigPath(const std::string &configPath) {
+    auto config = marian::bergamot::parseOptionsFromFilePath(configPath);
+    return service_.createCompatibleModel(config);
+  }
+
+  std::vector<Response> translate(Model model, std::vector<std::string> &inputs,
+                                  const ResponseOptions &options) {
+    py::scoped_ostream_redirect outstream(
+        std::cout,                                // std::ostream&
+        py::module_::import("sys").attr("stdout") // Python output
+    );
+    py::scoped_ostream_redirect errstream(
+        std::cerr,                                // std::ostream&
+        py::module_::import("sys").attr("stderr") // Python output
+    );
+
+    py::call_guard<py::gil_scoped_release> gil_guard;
+
+    // Prepare promises, save respective futures. Have callback's in async set
+    // value to the promises.
+    std::vector<std::future<Response>> futures;
+    std::vector<std::promise<Response>> promises;
+    promises.resize(inputs.size());
+
+    for (size_t i = 0; i < inputs.size(); i++) {
+      auto callback = [&promises, i](Response &&response) {
+        promises[i].set_value(std::move(response));
+      };
+
+      service_.translate(model, std::move(inputs[i]), std::move(callback),
+                         options);
+
+      futures.push_back(std::move(promises[i].get_future()));
+    }
+
+    // Wait on all futures to be ready.
+    std::vector<Response> responses;
+    for (size_t i = 0; i < futures.size(); i++) {
+      futures[i].wait();
+      responses.push_back(std::move(futures[i].get()));
+    }
+
+    return responses;
+  }
+
+  std::vector<Response> pivot(Model first, Model second,
```

Comment:
This function should be elsewhere, it is not a python specific thin interface.

Reply:
The original (async) API decided on to adjust to translateLocally. WebAssembly thing is single-threaded. I quickly notice that the GIL is not released and picked up again here, which I should fix. Other than that this function is unique to python, the moment GIL code comes in. Might even want to redirect stdout/stderr here in prep.

Reply:
I think I agree with Jerin here. There is (and probably will be more) python-specific stuff sprinkled through it to control the GIL and type conversions. It still delegates the bulk of the work to `AsyncService::translate()` and `AsyncService::pivot()`.